File: apps/windows/TeXLocal.Core/Outline.cs
using System.Text.RegularExpressions;

namespace TeXLocal;

public sealed record OutlineItem(int Level, string Title, int Line);

/// <summary>A heading and the headings it encloses.</summary>
public sealed record OutlineNode(OutlineItem Item, IReadOnlyList<OutlineNode> Children);

/// <summary>A document's outline, words and lines, read in one pass.</summary>
public sealed record DocumentStats(IReadOnlyList<OutlineItem> Outline, int Words, int Lines);

/// <summary>
/// Sectioning commands in a document, as the browser version's outline reads
/// them (web/src/state.js SECTION_RE), and the word count read alongside.
/// </summary>
public static partial class Outline
{
    private static readonly string[] Levels = ["part", "chapter", "section", "subsection", "subsubsection", "paragraph"];

    [GeneratedRegex(@"\\(part|chapter|section|subsection|subsubsection|paragraph)\*?\s*(?:\[[^\]]*\])?\s*\{([^}]*)\}")]
    private static partial Regex Section();

    public static IReadOnlyList<OutlineItem> Parse(string text) => Analyze(text).Outline;

    /// <summary>
    /// The outline, words and lines (web/src/state.js analyzeDoc). Lines
    /// break where CodeMirror breaks them — at CR LF, CR or LF — so the line
    /// count is the editor's.
    /// </summary>
    public static DocumentStats Analyze(string text)
    {
        var items = new List<OutlineItem>();
        var words = 0;
        var lines = text.Split(["\r\n", "\r", "\n"], StringSplitOptions.None);
        for (var i = 0; i < lines.Length; i++)
        {
            var line = lines[i];
            if (IsComment(line))
            {
                continue;
            }
            var match = Section().Match(line);
            if (match.Success)
            {
                var title = match.Groups[2].Value;
                items.Add(new OutlineItem(
                    Array.IndexOf(Levels, match.Groups[1].Value),
                    title.Length == 0 ? "(untitled)" : title,
                    i + 1));
            }
            words += LineWords(line);
        }
        return new DocumentStats(items, words, lines.Length);
    }

    /// <summary>
    /// The headings that enclose a line, outermost first: the breadcrumb
    /// (web/src/state.js outlineChain).
    /// </summary>
    public static IReadOnlyList<OutlineItem> Chain(IReadOnlyList<OutlineItem> outline, int line)
    {
        var stack = new List<OutlineItem>();
        foreach (var item in outline)
        {
            if (item.Line > line)
            {
                break;
            }
            while (stack.Count > 0 && stack[^1].Level >= item.Level)
            {
                stack.RemoveAt(stack.Count - 1);
            }
            stack.Add(item);
        }
        return stack;
    }

    /// <summary>
    /// Each heading's depth in the document's actual nesting: how many
    /// headings enclose it. A subsection before any section sits flush,
    /// rather than under a parent that isn't there (apps/macos Outline.swift).
    /// </summary>
    public static IReadOnlyList<int> Depths(IReadOnlyList<OutlineItem> outline)
    {
        var stack = new List<int>();
        var depths = new List<int>(outline.Count);
        foreach (var item in outline)
        {
            while (stack.Count > 0 && stack[^1] >= item.Level)
            {
                stack.RemoveAt(stack.Count - 1);
            }
            stack.Add(item.Level);
            depths.Add(stack.Count - 1);
        }
        return depths;
    }

    /// <summary>The outline as a tree by how the headings nest, for a sidebar with expanders.</summary>
    public static IReadOnlyList<OutlineNode> Tree(IReadOnlyList<OutlineItem> outline)
    {
        var depths = Depths(outline);
        var index = 0;
        List<OutlineNode> Children(int depth)
        {
            var nodes = new List<OutlineNode>();
            while (index < outline.Count && depths[index] == depth)
            {
                var item = outline[index++];
                nodes.Add(new OutlineNode(item, Children(depth + 1)));
            }
            return nodes;
        }
        return Children(0);
    }

    /// <summary>An empty heading by its kind — "Untitled subsection" — where the web writes "(untitled)".</summary>
    public static string DisplayTitle(OutlineItem item)
    {
        if (item.Title != "(untitled)")
        {
            return item.Title;
        }
        string[] kinds = ["part", "chapter", "section", "subsection", "subsubsection", "paragraph"];
        return "Untitled " + (item.Level >= 0 && item.Level < kinds.Length ? kinds[item.Level] : "section");
    }

    // ---------- word count ----------

    // The rules below are web/src/state.js lineWords, whose regular
    // expressions run on JavaScript's terms; they are spelled out here so
    // every line counts the same in both (as apps/macos Outline.swift does).
    // Every character involved is in the BMP, so UTF-16 units serve.

    /// <summary>JavaScript's \s.</summary>
    private static bool IsSpace(char c) =>
        c is '\t' or '\n' or '\u000B' or '\u000C' or '\r' or ' ' or '\u00A0' or '\u1680'
            or (>= '\u2000' and <= '\u200A') or '\u2028' or '\u2029' or '\u202F' or '\u205F' or '\u3000' or '\uFEFF';

    /// <summary>TeX's special characters, which separate words like spaces do.</summary>
    private static bool IsSpecial(char c) => c is '{' or '}' or '$' or '&' or '_' or '^' or '~' or '\\' or '%';

    private static bool IsComment(string line)
    {
        foreach (var c in line)
        {
            if (!IsSpace(c))
            {
                return c == '%';
            }
        }
        return false;
    }

    /// <summary>
    /// Rough word count of a prose line: drop the comment, then commands with
    /// a star and one [argument], then TeX's special characters, and count
    /// the runs left that contain a letter.
    /// </summary>
    public static int LineWords(string line)
    {
        var text = CommentStart(line) is { } comment ? line[..comment] : line;
        var words = 0;
        var letter = false;
        var i = 0;
        // One step past the end, as a space, closes the last run.
        while (i <= text.Length)
        {
            var c = i < text.Length ? text[i] : ' ';
            if (c == '\\' && i + 1 < text.Length && char.IsAsciiLetter(text[i + 1]))
            {
                i++;
                while (i < text.Length && char.IsAsciiLetter(text[i]))
                {
                    i++;
                }
                if (i < text.Length && text[i] == '*')
                {
                    i++;
                }
                if (i < text.Length && text[i] == '[' && text.IndexOf(']', i + 1) is var close and >= 0)
                {
                    i = close + 1;
                }
                if (letter)
                {
                    words++;
                }
                letter = false;
                continue;
            }
            if (IsSpace(c) || IsSpecial(c))
            {
                if (letter)
                {
                    words++;
                }
                letter = false;
            }
            else if (char.IsAsciiLetter(c) || c is >= '\u00C0' and <= '\u017E')
            {
                letter = true;
            }
            i++;
        }
        return words;
    }

    /// <summary>
    /// The first % no backslash escapes. JavaScript's . stops at U+2028 and
    /// U+2029, so a % with either after it starts no comment there.
    /// </summary>
    private static int? CommentStart(string text)
    {
        var from = text.LastIndexOfAny(['\u2028', '\u2029']) + 1;
        for (var i = from; i < text.Length; i++)
        {
            if (text[i] == '%' && (i == 0 || text[i - 1] != '\\'))
            {
                return i;
            }
        }
        return null;
    }
}

File: apps/macos/TeXLocal/Outline.swift
import Foundation

struct OutlineItem: Identifiable, Hashable {
    let id: Int
    let level: Int
    let title: String
    let line: Int
}

/// Sectioning commands in a document, as the browser version's outline reads
/// them (web/src/state.js `SECTION_RE`), and the word count read alongside.
enum Outline {
    private static let levels = ["part", "chapter", "section", "subsection", "subsubsection", "paragraph"]

    static func parse(_ text: String) -> [OutlineItem] {
        analyze(text).outline
    }

    /// The outline, words and lines in one pass (web/src/state.js
    /// `analyzeDoc`). Lines break where CodeMirror breaks them — at CR LF, CR
    /// or LF — so the line count is the editor's.
    static func analyze(_ text: String) -> (outline: [OutlineItem], words: Int, lines: Int) {
        // Local rather than a static: Regex is not Sendable.
        let pattern = /\\(part|chapter|section|subsection|subsubsection|paragraph)\*?\s*(?:\[[^\]]*\])?\s*\{([^}]*)\}/
        var items: [OutlineItem] = []
        var words = 0
        var lines = 0
        for line in text.split(omittingEmptySubsequences: false, whereSeparator: { $0 == "\n" || $0 == "\r" || $0 == "\r\n" }) {
            lines += 1
            if isComment(line) { continue }
            if let m = line.firstMatch(of: pattern) {
                items.append(OutlineItem(
                    id: items.count,
                    level: levels.firstIndex(of: String(m.1)) ?? 2,
                    title: m.2.isEmpty ? "(untitled)" : String(m.2),
                    line: lines
                ))
            }
            words += lineWords(line)
        }
        return (items, words, lines)
    }

    /// Each heading's depth in the document's actual nesting: how many
    /// headings enclose it. A subsection before any section sits flush,
    /// rather than indented under a parent that isn't there.
    static func depths(_ outline: [OutlineItem]) -> [Int] {
        var stack: [Int] = []
        return outline.map { item in
            while let last = stack.last, last >= item.level { stack.removeLast() }
            stack.append(item.level)
            return stack.count - 1
        }
    }

    /// An empty heading by its kind — "Untitled Subsection" — where the web
    /// writes "(untitled)".
    static func displayTitle(_ item: OutlineItem) -> String {
        guard item.title == "(untitled)" else { return item.title }
        let kinds = ["Part", "Chapter", "Section", "Subsection", "Subsubsection", "Paragraph"]
        return "Untitled " + (kinds.indices.contains(item.level) ? kinds[item.level] : "Section")
    }

    /// The headings that enclose a line, outermost first: the breadcrumb
    /// (web/src/state.js `outlineChain`).
    static func chain(_ outline: [OutlineItem], at line: Int) -> [OutlineItem] {
        var stack: [OutlineItem] = []
        for item in outline {
            if item.line > line { break }
            while let last = stack.last, last.level >= item.level { stack.removeLast() }
            stack.append(item)
        }
        return stack
    }

    // ---------- word count ----------

    // The rules below are web/src/state.js `lineWords`, whose regular
    // expressions run on JavaScript's terms; they are spelled out over Unicode
    // scalars here so that every line counts the same in both.

    /// JavaScript's `\s`.
    private static let whitespace = Set("\t\n\u{0B}\u{0C}\r \u{A0}\u{1680}\u{2000}\u{2001}\u{2002}\u{2003}\u{2004}\u{2005}\u{2006}\u{2007}\u{2008}\u{2009}\u{200A}\u{2028}\u{2029}\u{202F}\u{205F}\u{3000}\u{FEFF}".unicodeScalars)
    /// TeX's special characters, which separate words like spaces do.
    private static let special = Set("{}$&_^~\\%".unicodeScalars)

    /// A line whose first non-space character is `%`.
    private static func isComment(_ line: Substring) -> Bool {
        line.unicodeScalars.first(where: { !whitespace.contains($0) }) == "%"
    }

    /// Rough word count of a prose line: drop the comment, then commands with
    /// a star and one [argument], then TeX's special characters, and count the
    /// runs left that contain a letter.
    static func lineWords(_ line: Substring) -> Int {
        var text = Array(line.unicodeScalars)
        if let comment = commentStart(text) { text.removeSubrange(comment...) }
        var words = 0
        var letter = false
        var i = 0
        // One step past the end, as a space, closes the last run.
        while i <= text.count {
            let c: Unicode.Scalar = i < text.count ? text[i] : " "
            if c == "\\", i + 1 < text.count, isASCIILetter(text[i + 1]) {
                i += 1
                while i < text.count, isASCIILetter(text[i]) { i += 1 }
                if i < text.count, text[i] == "*" { i += 1 }
                if i < text.count, text[i] == "[", let close = text[(i + 1)...].firstIndex(of: "]") { i = close + 1 }
                if letter { words += 1 }
                letter = false
                continue
            }
            if whitespace.contains(c) || special.contains(c) {
                if letter { words += 1 }
                letter = false
            } else if isWordLetter(c) {
                letter = true
            }
            i += 1
        }
        return words
    }

    /// The first `%` no backslash escapes. JavaScript's `.` stops at U+2028
    /// and U+2029, so a `%` with either after it starts no comment there.
    private static func commentStart(_ text: [Unicode.Scalar]) -> Int? {
        let from = text.lastIndex(where: { $0 == "\u{2028}" || $0 == "\u{2029}" }).map { $0 + 1 } ?? 0
        return (from..<text.count).first(where: { i in
            text[i] == "%" && (i == 0 || text[i - 1] != "\\")
        })
    }

    private static func isASCIILetter(_ c: Unicode.Scalar) -> Bool {
        (0x41...0x5A).contains(c.value) || (0x61...0x7A).contains(c.value)
    }

    /// `[A-Za-zÀ-ž]`.
    private static func isWordLetter(_ c: Unicode.Scalar) -> Bool {
        isASCIILetter(c) || (0xC0...0x17E).contains(c.value)
    }
}
